File: packages/git-core/src/workflows/stash.ts
/**
 * Stash and tag workflows.
 *
 * Two lookups are the safety-critical part of the stash half:
 *
 * - **the locator is re-resolved before every write.** `stash@{0}` is a position in a
 *   reflog, and anyone's `git stash` — including the user's own terminal — shifts
 *   every position after it. Each workflow re-runs `rev-parse <locator>^{commit}`
 *   and compares the object name to the one the request carried; a mismatch refuses
 *   the operation instead of touching a different entry.
 * - **a conflicted pop is not a failed pop.** Git applies the stash, leaves the
 *   conflict in the working tree, and *does not drop* the entry. The workflow
 *   re-checks that the stash still resolves and reports `conflict` with
 *   `stashPreserved`, which the effect maps to `needsAttention`.
 *
 * Tag creation checks for an existing name first: refusing to overwrite is the
 * behaviour, and doing it as a pre-check produces a clear message instead of Git's
 * `already exists` after the object was prepared.
 */
import type { GitCommandSpec } from "../ports.js";
import {
  boundedDiagnostic,
  GitWorkflowError,
  runRequired,
  type GitEngine,
  type GitFailureCode,
} from "./engine.js";
import {
  planStashApply,
  planStashDrop,
  planStashPop,
  planStashPush,
  planStashResolve,
} from "../plan/commit.js";
import {
  planTagCreateAnnotated,
  planTagCreateLightweight,
  planTagDelete,
  planTagExists,
} from "../plan/tags.js";

export type SimpleOutcome =
  | { readonly kind: "done" }
  | {
      readonly kind: "refused";
      readonly code: GitFailureCode;
      readonly exitCode: number | null;
      readonly diagnostic: string;
    };

function refusalOf(
  error: unknown,
): Extract<SimpleOutcome, { kind: "refused" }> | null {
  if (error instanceof GitWorkflowError) {
    return {
      kind: "refused",
      code: error.code,
      exitCode: error.exitCode,
      diagnostic: error.diagnostic,
    };
  }
  return null;
}

async function runSimple(
  engine: GitEngine,
  spec: GitCommandSpec,
): Promise<SimpleOutcome> {
  try {
    await runRequired(engine, spec);
    return { kind: "done" };
  } catch (error) {
    const refusal = refusalOf(error);
    if (refusal !== null) {
      return refusal;
    }
    throw error;
  }
}

/** Create a stash; the message (if any) travels in argv as `--message=`. */
export function createStash(
  engine: GitEngine,
  context: { readonly cwdHandle: string },
  input: {
    readonly message: string | null;
    readonly includeUntracked: boolean;
    readonly keepIndex: boolean;
  },
): Promise<SimpleOutcome> {
  return runSimple(engine, planStashPush(context, input));
}

export type LocatorCheck =
  | { readonly kind: "matches" }
  | { readonly kind: "moved"; readonly resolvedOid: string | null }
  | { readonly kind: "unresolvable" };

/** Resolve the locator and compare it to the object the request named. */
export async function checkStashLocator(
  engine: GitEngine,
  context: { readonly cwdHandle: string },
  input: { readonly locator: string; readonly oid: string },
): Promise<LocatorCheck> {
  const result = await engine.run(planStashResolve(context, input.locator));
  if (result.termination !== "exit" || result.exitCode !== 0) {
    return { kind: "unresolvable" };
  }
  let text = "";
  for (const byte of result.stdout) {
    if (byte === 0x0a || byte === 0x0d) {
      break;
    }
    text += String.fromCharCode(byte);
  }
  return text.trim() === input.oid
    ? { kind: "matches" }
    : { kind: "moved", resolvedOid: text.trim() === "" ? null : text.trim() };
}

export type PopOutcome =
  | { readonly kind: "done" }
  | {
      readonly kind: "conflict";
      readonly diagnostic: string;
      /** True when the stash entry is still present after the conflict. */
      readonly stashPreserved: boolean;
    }
  | {
      readonly kind: "refused";
      readonly code: GitFailureCode;
      readonly exitCode: number | null;
      readonly diagnostic: string;
    };

/**
 * Pop a stash: apply, and drop only when Git applied it cleanly.
 *
 * On a non-zero exit the stash is re-checked by object name: Git's own rule is that
 * a conflicted pop leaves it in place, and the re-check is what makes the report
 * evidence rather than a belief about Git's behaviour.
 */
export async function popStash(
  engine: GitEngine,
  context: { readonly cwdHandle: string },
  input: {
    readonly locator: string;
    readonly oid: string;
    readonly restoreIndex: boolean;
  },
): Promise<PopOutcome> {
  const spec = planStashPop(context, {
    locator: input.locator,
    restoreIndex: input.restoreIndex,
  });
  const result = await engine.run(spec);
  if (result.termination === "exit" && result.exitCode === 0) {
    return { kind: "done" };
  }
  const diagnostic = boundedDiagnostic(result.stderr);
  if (result.termination !== "exit") {
    // A timeout or a signal: whether the stash was applied (or dropped) is not
    // known, and the caller must not continue to any further step.
    return {
      kind: "refused",
      code:
        result.termination === "timeout"
          ? "GitTimedOut"
          : result.termination === "signal"
            ? "GitTerminatedBySignal"
            : "GitNotStarted",
      exitCode: result.exitCode,
      diagnostic,
    };
  }
  const after = await checkStashLocator(engine, context, {
    locator: input.locator,
    oid: input.oid,
  });
  return {
    kind: "conflict",
    diagnostic:
      diagnostic.trim().length > 0
        ? diagnostic
        : "the pop did not apply cleanly",
    stashPreserved: after.kind === "matches",
  };
}

export function applyStash(
  engine: GitEngine,
  context: { readonly cwdHandle: string },
  input: { readonly locator: string; readonly restoreIndex: boolean },
): Promise<SimpleOutcome> {
  return runSimple(
    engine,
    planStashApply(context, {
      locator: input.locator,
      restoreIndex: input.restoreIndex,
    }),
  );
}

export function dropStash(
  engine: GitEngine,
  context: { readonly cwdHandle: string },
  locator: string,
): Promise<SimpleOutcome> {
  return runSimple(engine, planStashDrop(context, locator));
}

export type TagCreateOutcome =
  | { readonly kind: "created" }
  | { readonly kind: "exists" }
  | {
      readonly kind: "refused";
      readonly code: GitFailureCode;
      readonly exitCode: number | null;
      readonly diagnostic: string;
    };

/** Create a tag, refusing to overwrite an existing name. */
export async function createTag(
  engine: GitEngine,
  context: { readonly cwdHandle: string },
  input: {
    readonly name: string;
    readonly targetOid: string | null;
    readonly annotation: { readonly message: Uint8Array } | null;
  },
): Promise<TagCreateOutcome> {
  const existing = await engine.run(planTagExists(context, input.name));
  if (existing.termination === "exit" && existing.exitCode === 0) {
    return { kind: "exists" };
  }
  if (existing.termination !== "exit" || existing.exitCode !== 1) {
    // Anything other than the "no such tag" exit is a real failure to look it up,
    // and creating on top of an unreadable state is not safe.
    return {
      kind: "refused",
      code: "GitCommandFailed",
      exitCode: existing.exitCode,
      diagnostic: boundedDiagnostic(existing.stderr),
    };
  }
  const spec =
    input.annotation === null
      ? planTagCreateLightweight(context, {
          name: input.name,
          targetOid: input.targetOid,
        })
      : planTagCreateAnnotated(context, {
          name: input.name,
          targetOid: input.targetOid,
          message: input.annotation.message,
        });
  const created = await runSimple(engine, spec);
  return created.kind === "done" ? { kind: "created" } : created;
}

export function deleteTag(
  engine: GitEngine,
  context: { readonly cwdHandle: string },
  name: string,
): Promise<SimpleOutcome> {
  return runSimple(engine, planTagDelete(context, name));
}

File: crates/refyard-host/src/writes/stash.rs
//! Stash operations: create, apply, pop and drop one entry.
//!
//! The safety-critical rule is that `stash@{n}` is a *position in a reflog*, not a
//! stable identifier. Anyone's `git stash` — including the user's own terminal —
//! shifts every position after it. So apply, pop and drop first re-resolve the
//! locator through `rev-parse <locator>^{commit}` and compare the object name to the
//! one the request carried; a mismatch is a stale request and refuses rather than
//! acting on a different entry. `pop` and `drop` are destructive and gate on an
//! explicit `confirmed`, and `pop` drops only when Git applied cleanly — a conflicted
//! pop leaves the entry in place, which this module reports as `needsAttention` (work
//! for a human) rather than a failure that would suggest nothing happened.

use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;

use refyard_contract::problem::{Problem, ProblemCode};
use refyard_contract::reads::MutationKind;
use refyard_core::outcome::{ExecutionState, RunOutcome};

use crate::jobs::journal::EffectOutcome;
use crate::jobs::{EffectRequest, MutationEffect, MutationOperation};

use super::{
    clean_exit, conflicted_count, diagnostic_of, invalid_payload, refusal_problem, refused,
    require_confirmed, result_of, unknown_problem, wrong_payload, WriteHost, WriteTarget,
};

/// `git stash push` — save the working-tree changes.
pub struct CreateStashEffect {
    pub(super) host: Arc<WriteHost>,
}

/// `git stash apply` — restore an entry and keep it.
pub struct ApplyStashEffect {
    pub(super) host: Arc<WriteHost>,
}

/// `git stash pop` — restore and drop, but only on a clean apply.
pub struct PopStashEffect {
    pub(super) host: Arc<WriteHost>,
}

/// `git stash drop` — discard one entry. Destructive.
pub struct DropStashEffect {
    pub(super) host: Arc<WriteHost>,
}

impl MutationEffect for CreateStashEffect {
    fn kind(&self) -> MutationKind {
        MutationKind::CreateStash
    }
    fn run<'a>(
        &'a self,
        request: EffectRequest<'a>,
    ) -> Pin<Box<dyn Future<Output = EffectOutcome> + Send + 'a>> {
        let host = Arc::clone(&self.host);
        Box::pin(async move { create_stash(&host, &request).await })
    }
}

impl MutationEffect for ApplyStashEffect {
    fn kind(&self) -> MutationKind {
        MutationKind::ApplyStash
    }
    fn run<'a>(
        &'a self,
        request: EffectRequest<'a>,
    ) -> Pin<Box<dyn Future<Output = EffectOutcome> + Send + 'a>> {
        let host = Arc::clone(&self.host);
        Box::pin(async move { apply_stash(&host, &request).await })
    }
}

impl MutationEffect for PopStashEffect {
    fn kind(&self) -> MutationKind {
        MutationKind::PopStash
    }
    fn run<'a>(
        &'a self,
        request: EffectRequest<'a>,
    ) -> Pin<Box<dyn Future<Output = EffectOutcome> + Send + 'a>> {
        let host = Arc::clone(&self.host);
        Box::pin(async move { pop_stash(&host, &request).await })
    }
}

impl MutationEffect for DropStashEffect {
    fn kind(&self) -> MutationKind {
        MutationKind::DropStash
    }
    fn run<'a>(
        &'a self,
        request: EffectRequest<'a>,
    ) -> Pin<Box<dyn Future<Output = EffectOutcome> + Send + 'a>> {
        let host = Arc::clone(&self.host);
        Box::pin(async move { drop_stash(&host, &request).await })
    }
}

/// Where a stash locator points now, versus the object the request named.
enum LocatorCheck {
    Matches,
    Moved,
    Unresolvable,
}

/// Re-resolve the locator and compare it to the object the request carried.
async fn check_stash_locator(target: &WriteTarget, locator: &str, oid: &str) -> LocatorCheck {
    let Ok(plan) = refyard_core::plan::commit::plan_stash_resolve(locator) else {
        return LocatorCheck::Unresolvable;
    };
    let Ok(outcome) = target
        .executor
        .try_run(
            target.record.location.canonical_worktree.as_str(),
            &plan,
            None,
        )
        .await
    else {
        return LocatorCheck::Unresolvable;
    };
    if !clean_exit(&outcome) {
        return LocatorCheck::Unresolvable;
    }
    // The object name is ASCII hex plus a newline; byte-level extraction is exact.
    let text = String::from_utf8(outcome.stdout).unwrap_or_default();
    let resolved = text.split(['\n', '\r']).next().unwrap_or("").trim();
    if resolved == oid {
        LocatorCheck::Matches
    } else {
        LocatorCheck::Moved
    }
}

/// The refusal a locator mismatch carries: a stale request, never a different entry.
fn locator_problem(operation_id: &str, check: LocatorCheck) -> Option<EffectOutcome> {
    match check {
        LocatorCheck::Matches => None,
        LocatorCheck::Unresolvable => Some(refused(
            operation_id,
            Problem::new(
                ProblemCode::NotFound,
                "that stash entry no longer exists; reload the stash list",
            ),
        )),
        LocatorCheck::Moved => Some(refused(
            operation_id,
            Problem::new(
                ProblemCode::Conflict,
                "the stash list moved since this request was prepared, and the locator now points at a different entry; reload and select the stash again",
            ),
        )),
    }
}

/// Reads back the current head so a success can name it (stash leaves HEAD alone, so
/// this is the same oid the request was prepared against — reported as evidence, not
/// as a claim that anything moved).
async fn head_oid(target: &WriteTarget) -> Option<String> {
    probe_head(target).await
}

async fn probe_head(target: &WriteTarget) -> Option<String> {
    let plan = refyard_core::plan::GitPlan::read(vec!["rev-parse".to_string(), "HEAD".to_string()]);
    let outcome = target
        .executor
        .try_run(
            target.record.location.canonical_worktree.as_str(),
            &plan,
            None,
        )
        .await
        .ok()?;
    if !clean_exit(&outcome) {
        return None;
    }
    let text = String::from_utf8(outcome.stdout).unwrap_or_default();
    let trimmed = text.trim();
    if trimmed.is_empty() {
        None
    } else {
        Some(trimmed.to_string())
    }
}

/// A started-but-not-clean run's outcome, without a conflict case: a clean non-zero
/// is a refusal, and a timeout, signal or truncated stream is `unknown` because the
/// change may be half-applied and is never retried.
fn started_failure(
    operation_id: &str,
    command: &'static str,
    outcome: &RunOutcome,
) -> EffectOutcome {
    let finished = outcome.state == ExecutionState::Completed && outcome.output_complete;
    if finished {
        refused(operation_id, refusal_problem(command, outcome))
    } else {
        EffectOutcome::Unknown {
            reason: format!("{command} did not finish cleanly"),
            problem: unknown_problem(
                command,
                "the command did not finish cleanly; whether it changed anything is not known, and nothing was retried",
            )
            .for_operation(operation_id.to_string()),
        }
    }
}

fn succeeded(summary: String, new_head_oid: Option<String>) -> EffectOutcome {
    EffectOutcome::Succeeded {
        result: result_of(summary, None, new_head_oid),
    }
}

fn needs_attention(operation_id: &str, message: String) -> EffectOutcome {
    EffectOutcome::NeedsAttention {
        problem: Problem::new(ProblemCode::NeedsAttention, message)
            .for_operation(operation_id.to_string()),
    }
}

async fn create_stash(host: &Arc<WriteHost>, request: &EffectRequest<'_>) -> EffectOutcome {
    let operation_id = request.operation_id;
    let MutationOperation::CreateStash {
        message,
        include_untracked,
        keep_index,
    } = &request.request.operation
    else {
        return wrong_payload(operation_id, "createStash");
    };
    let plan = match refyard_core::plan::commit::plan_stash_push(
        message.as_deref(),
        *include_untracked,
        *keep_index,
    ) {
        Ok(plan) => plan,
        Err(error) => return invalid_payload(operation_id, error.to_string()),
    };
    let target = match host.resolve(request.request).await {
        Ok(target) => target,
        Err(problem) => return refused(operation_id, problem),
    };
    let outcome = match target
        .executor
        .try_run(
            target.record.location.canonical_worktree.as_str(),
            &plan,
            None,
        )
        .await
    {
        Ok(outcome) => outcome,
        Err(problem) => return refused(operation_id, problem),
    };
    if outcome.state == ExecutionState::NotStarted {
        return refused(operation_id, refusal_problem("git stash push", &outcome));
    }
    if clean_exit(&outcome) {
        let head = head_oid(&target).await;
        let label = message
            .as_deref()
            .and_then(|m| m.split('\n').next())
            .map(|first| format!(" ({first})"))
            .unwrap_or_default();
        return succeeded(format!("created a stash{label}"), head);
    }
    started_failure(operation_id, "git stash push", &outcome)
}

async fn apply_stash(host: &Arc<WriteHost>, request: &EffectRequest<'_>) -> EffectOutcome {
    let operation_id = request.operation_id;
    let MutationOperation::ApplyStash {
        stash,
        restore_index,
    } = &request.request.operation
    else {
        return wrong_payload(operation_id, "applyStash");
    };
    let plan = match refyard_core::plan::commit::plan_stash_apply(&stash.locator, *restore_index) {
        Ok(plan) => plan,
        Err(error) => return invalid_payload(operation_id, error.to_string()),
    };
    let target = match host.resolve(request.request).await {
        Ok(target) => target,
        Err(problem) => return refused(operation_id, problem),
    };
    if let Some(mismatch) = locator_problem(
        operation_id,
        check_stash_locator(&target, &stash.locator, &stash.oid).await,
    ) {
        return mismatch;
    }
    let outcome = match target
        .executor
        .try_run(
            target.record.location.canonical_worktree.as_str(),
            &plan,
            None,
        )
        .await
    {
        Ok(outcome) => outcome,
        Err(problem) => return refused(operation_id, problem),
    };
    if outcome.state == ExecutionState::NotStarted {
        return refused(operation_id, refusal_problem("git stash apply", &outcome));
    }
    if clean_exit(&outcome) {
        let head = head_oid(&target).await;
        return succeeded(
            format!("applied {}; the stash is kept", stash.locator),
            head,
        );
    }
    // A clean non-zero: a conflicted apply leaves conflicts for a human and keeps the
    // entry; anything else is Git refusing outright.
    let finished = outcome.state == ExecutionState::Completed && outcome.output_complete;
    if finished {
        let conflicts = conflicted_count(&target).await.unwrap_or(0);
        if conflicts > 0 {
            let plural = if conflicts == 1 { "path" } else { "paths" };
            return needs_attention(
                operation_id,
                format!(
                    "the apply left conflicts in the working tree and the stash is kept (nothing was dropped): {conflicts} conflicted {plural}: {}",
                    diagnostic_of(&outcome)
                ),
            );
        }
    }
    started_failure(operation_id, "git stash apply", &outcome)
}

async fn pop_stash(host: &Arc<WriteHost>, request: &EffectRequest<'_>) -> EffectOutcome {
    let operation_id = request.operation_id;
    let MutationOperation::PopStash {
        stash,
        restore_index,
        confirmed,
    } = &request.request.operation
    else {
        return wrong_payload(operation_id, "popStash");
    };
    if let Err(outcome) = require_confirmed(operation_id, *confirmed, "pop a stash") {
        return outcome;
    }
    let plan = match refyard_core::plan::commit::plan_stash_pop(&stash.locator, *restore_index) {
        Ok(plan) => plan,
        Err(error) => return invalid_payload(operation_id, error.to_string()),
    };
    let target = match host.resolve(request.request).await {
        Ok(target) => target,
        Err(problem) => return refused(operation_id, problem),
    };
    if let Some(mismatch) = locator_problem(
        operation_id,
        check_stash_locator(&target, &stash.locator, &stash.oid).await,
    ) {
        return mismatch;
    }
    let outcome = match target
        .executor
        .try_run(
            target.record.location.canonical_worktree.as_str(),
            &plan,
            None,
        )
        .await
    {
        Ok(outcome) => outcome,
        Err(problem) => return refused(operation_id, problem),
    };
    if outcome.state == ExecutionState::NotStarted {
        return refused(operation_id, refusal_problem("git stash pop", &outcome));
    }
    if clean_exit(&outcome) {
        let head = head_oid(&target).await;
        return succeeded(
            format!("popped {}; the entry was dropped", stash.locator),
            head,
        );
    }
    let finished = outcome.state == ExecutionState::Completed && outcome.output_complete;
    if finished {
        let conflicts = conflicted_count(&target).await.unwrap_or(0);
        if conflicts > 0 {
            // Git's own rule is that a conflicted pop leaves the entry in place; the
            // re-check is what makes "still there" evidence rather than a belief.
            let preserved = matches!(
                check_stash_locator(&target, &stash.locator, &stash.oid).await,
                LocatorCheck::Matches
            );
            let diagnostic = diagnostic_of(&outcome);
            let message = if preserved {
                format!(
                    "the pop left conflicts in the working tree and the stash is still there (nothing was dropped): {diagnostic}"
                )
            } else {
                format!(
                    "the pop left conflicts and the stash could no longer be found afterwards — check the stash list before retrying: {diagnostic}"
                )
            };
            return needs_attention(operation_id, message);
        }
    }
    // A timeout or signal on a pop is unknown: whether it applied (or dropped) is not
    // known, and nothing is ever continued past it.
    started_failure(operation_id, "git stash pop", &outcome)
}

async fn drop_stash(host: &Arc<WriteHost>, request: &EffectRequest<'_>) -> EffectOutcome {
    let operation_id = request.operation_id;
    let MutationOperation::DropStash { stash, confirmed } = &request.request.operation else {
        return wrong_payload(operation_id, "dropStash");
    };
    if let Err(outcome) = require_confirmed(operation_id, *confirmed, "drop a stash") {
        return outcome;
    }
    let plan = match refyard_core::plan::commit::plan_stash_drop(&stash.locator) {
        Ok(plan) => plan,
        Err(error) => return invalid_payload(operation_id, error.to_string()),
    };
    let target = match host.resolve(request.request).await {
        Ok(target) => target,
        Err(problem) => return refused(operation_id, problem),
    };
    if let Some(mismatch) = locator_problem(
        operation_id,
        check_stash_locator(&target, &stash.locator, &stash.oid).await,
    ) {
        return mismatch;
    }
    let outcome = match target
        .executor
        .try_run(
            target.record.location.canonical_worktree.as_str(),
            &plan,
            None,
        )
        .await
    {
        Ok(outcome) => outcome,
        Err(problem) => return refused(operation_id, problem),
    };
    if outcome.state == ExecutionState::NotStarted {
        return refused(operation_id, refusal_problem("git stash drop", &outcome));
    }
    if clean_exit(&outcome) {
        let head = head_oid(&target).await;
        return succeeded(format!("dropped {}", stash.locator), head);
    }
    started_failure(operation_id, "git stash drop", &outcome)
}
